bpo-42213: Remove redundant cyclic GC hack in sqlite3

erlend-aasland opened the pull request:
The sqlite3 module now fully implements the GC protocol; there's no
need for this workaround anymore.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- issue-number: [bpo-42213](https://bugs.python.org/issue42213) -->
https://bugs.python.org/issue42213
<!-- /issue-number -->

Commit:
bpo-42213: Remove redundant cyclic GC hack in sqlite3

The sqlite3 module now fully implements the GC protocol, so there's no
need for this workaround anymore.

pablogsal commented on Lib/sqlite3/test/dbapi.py:
```diff
@@ -190,6 +193,8 @@ def test_open_uri(self):
             with self.assertRaises(sqlite.OperationalError):
                 cx.execute('insert into test(id) values(1)')
+        cx.close()
+        gc_collect()
```
Why the falls to GC collect here? These should not be necessary

erlend-aasland replied:
I know; it's the same issue as in #24203. I'll try my best to reproduce this on my Mac.

erlend-aasland replied:
FYI, discovered #26475 while debugging this, but it does unfortunately not fix this issue.

Commit:
Split open tests into their own test case, and cleanup more explicitly

Commit:
clean up test_trace_callback_content explicitly

Commit:
Document workarounds

Commit:
Use test.support.gc_collect iso. gc.collect

vstinner commented on Lib/sqlite3/test/dbapi.py:
```diff
@@ -170,10 +171,16 @@ def test_in_transaction_ro(self):
+class OpenTests(unittest.TestCase):
+    def tearDown(self):
+        # bpo-42213: ensure that TESTFN is closed before unlinking
+        gc_collect()
```
I am against a PR which calls gc.collect() to ensure that a resource is released. It doesn't look reliable at all. It sounds like a way to hide a real bug, and it will likely bite us later.

erlend-aasland replied:
I agree; I'll revert the workaround.

Commit:
Revert gc workarounds

Reverts commits:
- 866e226c87558dd26e3ba309b8f5c801c2366734.
- 5ff1a806d22eec8ad09013108e2de76f0d9a55c9.
- 1e3f5898ac9a29da16b0713f3a344fbadea3c915.
- 6cb435b8dacd39b3e0dc04b2280b5ddfb48e4af9.

Commit:
Address review: Explicitly close opened database files

Commit:
Try to harden connection close

- add wrapper for sqlite3_close_v2()
- explicitly free pending statements before close()

Commit:
sqlite3_close_v2() always returns SQLITE_OK

pablogsal commented on Modules/_sqlite/connection.c:
```diff
@@ -353,23 +353,18 @@ static PyObject *
+         * up cursors, as they may have strong refs to statements. */
+        Py_CLEAR(self->statement_cache);
+        Py_CLEAR(self->statements);
+        Py_CLEAR(self->cursors);
```
What happens if you use the connection after you call `close()`? Are we checking for NULL for these veriables everywhere else? (Cannot check myself as I'm on my phone)

erlend-aasland replied:
You'll get a `sqlite3.ProgrammingError`. There's a lot of sanity checks in the code for this. The sqlite3 test suite runs fine, but of course, we do not have 100% code coverage yet. (Digression: see [issue 43553](https://bugs.python.org/issue43553) for improving `sqlite3` code coverage.)

After close, if you try to use a cursor, fetch from a pending statement, or manipulate the connection, you'll get a `sqlite3.ProgrammingError`:
```
>>> cx = sqlite3.connect(":memory:")
>>> cu = cx.cursor()
>>> res = cu.execute("select 1")
>>> cx.close()
>>> res.fetchall()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
sqlite3.ProgrammingError: Cannot operate on a closed database.
>>> cu.execute("select 1")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
sqlite3.ProgrammingError: Cannot operate on a closed database.
>>> cx.create_function("test", 1, lambda x: x)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
sqlite3.ProgrammingError: Cannot operate on a closed database.
>>> 
```

erlend-aasland replied:
I added some more tests wrt. this: 1b23df16d0dfad14b3f401ccbeeb34b7cbf3b82e
Also, there's some regression tests that exercise operations on closed connections (and closed cursors).

Let me know if you want me to add more tests.

erlend-aasland replied:
Here's a strange observation:
I did a ref leak test on Windows yesterday, and it segfaulted on `test_bpo31770`. After some testing, I reverted the change highlighted in this conversation; I removed the three `Py_CLEAR`, added `pysqlite_do_all_statements(self, ACTION_FINALIZE, 1)` back, but kept `connection_close()`. That fixed the ref leak test on Windows, and the test suite still worked; it did not "leak" open test database files. So, for the new PR, I'll keep this modification. We'll see how the CI fares :)

erlend-aasland replied:
cc. @vstinner

erlend-aasland replied:
CI seems to be fine with this: https://github.com/erlend-aasland/cpython/pull/9

vstinner commented on Lib/sqlite3/test/dbapi.py:
```diff
@@ -170,26 +170,35 @@ def test_in_transaction_ro(self):
         path = Path()
         with sqlite.connect(path) as cx:
             cx.execute('create table test(id integer)')
+        cx.close()
```
Maybe a helper context manager would be helpful? It would close the DB and then remove the file. Something like:

```
@contextlib.contextmanager
def connect(path):
  conn = sqlite.connect(path)
  try:
    yield conn
  finally:
    conn.close()
    unlink(path)
```

I'm not sure if it's worth it if there are just a few users of such helper function. Your PR is also good like that.

erlend-aasland replied:
I'm not sure it's worth it, but I'll keep it in mind if we add more such users.

erlend-aasland replied:
I changed my mind; the helper is nice. I'll add it :)

erlend-aasland replied:
See aba8dd724618df1a3ab2cad90b03436cb9bc9fd8

vstinner commented on Lib/sqlite3/test/hooks.py:
```diff
@@ -259,6 +259,8 @@ def trace(statement):
         con2.execute("create table bar(x)")
         cur.execute(queries[1])
+        con1.close()
+        con2.close()
```
May use try/finally to ensure that the connection is closed if an assertion fails. Or use an helper function.

erlend-aasland replied:
Good point

erlend-aasland replied:
I added d31403eb9885eb4654741a84b28f5dcd7179acad.

vstinner reviewed (commented):
Hum. Can you try to split your PR in two parts? First PR to rewrite the C code, second PR to rewrite the tests.

Commit:
Add tests that exercise stuff against a closed database

Commit:
Check connection on __enter__

Commit:
Harden test clean up

Commit:
Add managed resource helper

...and sort imports
